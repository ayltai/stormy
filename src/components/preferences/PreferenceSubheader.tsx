import { Divider, ListSubheader, } from '@mui/material';
import React from 'react';

import { Label, } from '../Label';

export const PreferenceSubheader = ({
    title,
    divider,
} : {
    title    : string,
    divider? : boolean,
}) => (
    <React.Fragment>
        {divider && <Divider />}
        <ListSubheader role='heading'>
            <Label variant='caption'>{title}</Label>
        </ListSubheader>
    </React.Fragment>
);
