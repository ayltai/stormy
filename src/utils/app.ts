import path from 'path';
import url from 'url';

import { MENU_ICON_SIZE, } from '../Constants';

import { convertTemperature, handleError, } from './index';

const getElectron = () => {
    let electron : any;

    try {
        electron = window.require('electron') || require('electron');
    } catch (error) {
        try {
            electron = require('electron');
        } catch (e) {
            electron = {
                ipcRenderer : {
                    on   : (message : string, args? : any) => {},
                    send : (message : string, args? : any) => {},
                },
                shell       : {
                    openExternal : (uri : string) => {},
                },
            };
        }
    }

    return electron;
};

let appPath : string = __dirname;
let nativeDarkMode   = false;

if (process.env.JEST_WORKER_ID === undefined) {
    getElectron().ipcRenderer.send('get-app-path');
    getElectron().ipcRenderer.on('set-app-path', (event : any, data : string) => appPath = data);

    getElectron().ipcRenderer.send('get-dark-mode');
    getElectron().ipcRenderer.on('set-dark-mode', (event : any, data : boolean) => nativeDarkMode = data);
}

export const getLocale = () => Intl.DateTimeFormat().resolvedOptions().locale;

export const isDarkMode = () => nativeDarkMode;

export const openUrl = (uri : string) => getElectron().shell.openExternal(uri);

export const setAppAutoLaunch = (enabled : boolean) => getElectron().ipcRenderer.send('set-auto-launch', enabled);

const createIcon = ({
    icon,
    scale,
} : {
    icon  : string,
    scale : number,
}) => new Promise<string>(resolve => {
    const size = MENU_ICON_SIZE * scale;

    const canvas = document.createElement('canvas');
    canvas.width  = size;
    canvas.height = size;

    const context = canvas.getContext('2d');
    const image   = new Image(size, size);

    image.onload = () => {
        context!.scale(scale, scale);
        context!.drawImage(image, 0, 0, size, size, 0, 0, MENU_ICON_SIZE, MENU_ICON_SIZE);

        resolve(canvas.toDataURL());
    };

    image.src = process.env.NODE_ENV === 'development' ? path.join('img', 'dark', `${icon}.svg`) : url.pathToFileURL(path.join(appPath, 'build', 'img', 'dark', `${icon}.svg`)).toString();
});

export const updateMenu = ({
    temperature,
    icon,
} : {
    temperature : number,
    icon        : string,
}) => {
    Promise.all([
        createIcon({
            icon,
            scale : 1,
        }),
        createIcon({
            icon,
            scale : 2,
        }),
    ]).then(([ icon1, icon2, ]) => {
        getElectron().ipcRenderer.send('refresh', {
            temperature : convertTemperature({
                temperature,
                displayUnit    : true,
                fractionDigits : 0,
            }),
            icon        : icon1,
            icon2x      : icon2,
        });
    }).catch(handleError);
};

export const exit = () => getElectron().ipcRenderer.send('exit');
