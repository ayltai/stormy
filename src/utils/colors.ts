const extractColors = (rgb : string) => {
    const numbers = rgb.split(',');

    return [
        parseInt(numbers[0].substring(4), 10),
        parseInt(numbers[1], 10),
        parseInt(numbers[2], 10),
    ];
};

const toHex = (decimal : number) => {
    const hex = decimal.toString(16);

    return hex.length === 1 ? `0${hex}` : hex;
};

export const makeTransparent = (hexColorCode : string, opacity : number) => {
    if (hexColorCode.startsWith('rgb')) {
        const colors = extractColors(hexColorCode);

        return `#${toHex(colors[0])}${toHex(colors[1])}${toHex(colors[2])}${toHex(Math.round(255 * opacity))}`;
    }

    const hex = hexColorCode.replace('#', '');

    return `#${hex.substring(0, 2)}${hex.substring(2, 4)}${hex.substring(4, 6)}${toHex(Math.round(255 * opacity))}`;
};
