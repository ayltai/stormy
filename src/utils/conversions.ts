export const convertTemperature = ({
    temperature,
    unit           = 'metric',
    displayUnit    = false,
    fractionDigits = 1,
} : {
    temperature     : number,
    unit?           : 'metric' | 'imperial',
    displayUnit?    : boolean,
    fractionDigits? : number,
}) => {
    const convertedTemperature = unit === 'metric' ? temperature : (temperature * 9 / 5) + 32;

    return displayUnit ? `${(fractionDigits === 0 ? Math.round(convertedTemperature) : convertedTemperature).toFixed(fractionDigits)}Â°${unit === 'metric' ? 'C' : 'F'}` : convertedTemperature;
};

export const convertSpeed = ({
    value,
    unit           = 'metric',
    displayUnit    = false,
    fractionDigits = 1,
} : {
    value           : number,
    unit?           : 'metric' | 'imperial',
    displayUnit?    : boolean,
    fractionDigits? : number,
}) => {
    const convertedSpeed = unit === 'metric' ? value : value * 2.23693629;

    return displayUnit ? `${(fractionDigits === 0 ? Math.round(convertedSpeed) : convertedSpeed).toFixed(fractionDigits)}${unit === 'metric' ? 'km/h' : 'mph'}` : convertedSpeed;
};
