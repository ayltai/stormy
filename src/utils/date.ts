export const getPartOfDay = (date : number) : 'day' | 'night' => {
    const hours = new Date(date).getHours();
    return hours >= 6 && hours < 18 ? 'day' : 'night';
};
