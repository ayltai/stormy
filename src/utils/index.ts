export const handleError = console.error;

export { changeDarkMode, exit, getLocale, isDarkMode, openUrl, setAppAutoLaunch, updateMenu, } from './app';
export { makeTransparent, } from './colors';
export { convertSpeed, convertTemperature, } from './conversions';
export { getPartOfDay, } from './dates';
export { capitalize, } from './strings';
export { getAlphaForPrecipIntensity, } from './weather';
