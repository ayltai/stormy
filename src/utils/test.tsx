import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, PreloadedState, Store, } from '@reduxjs/toolkit';
import '@testing-library/jest-dom/extend-expect';
import { render, } from '@testing-library/react';
import React, { ReactNode, } from 'react';
import { HashRouter, } from 'react-router-dom';
import { Provider, } from 'react-redux';

import { BACKGROUND_BLUR, BACKGROUND_DARKEN, } from '../Constants';
import { Place, } from '../models';
import { settingsReducer, } from '../redux';
import { palette, } from '../styles';

export const DEFAULT_LOCATION : Place = {
    name      : 'Hong Kong International Airport, Sky Plaza Road, Chek Lap Kok',
    latitude  : 22.3080,
    longitude : 113.9185,
};

export const MOCK_WEATHER_RESPONSE = {
    'lat'             : 39.31,
    'lon'             : -74.5,
    'timezone'        : 'America/New_York',
    'timezone_offset' : -18000,
    'current'         : {
        'dt'         : 1646318698,
        'sunrise'    : 1646306882,
        'sunset'     : 1646347929,
        'temp'       : 282.21,
        'feels_like' : 278.41,
        'pressure'   : 1014,
        'humidity'   : 65,
        'dew_point'  : 275.99,
        'uvi'        : 2.55,
        'clouds'     : 40,
        'visibility' : 10000,
        'wind_speed' : 8.75,
        'wind_deg'   : 360,
        'wind_gust'  : 13.89,
        'weather'    : [
            {
                'id'          : 802,
                'main'        : 'Clouds',
                'description' : 'scattered clouds',
                'icon'        : '03d',
            },
        ],
    },
    'hourly'          : [
        {
            'dt'         : 1646316000,
            'temp'       : 281.94,
            'feels_like' : 278.49,
            'pressure'   : 1014,
            'humidity'   : 67,
            'dew_point'  : 276.16,
            'uvi'        : 1.49,
            'clouds'     : 52,
            'visibility' : 10000,
            'wind_speed' : 7.16,
            'wind_deg'   : 313,
            'wind_gust'  : 10.71,
            'weather'    : [
                {
                    'id'          : 803,
                    'main'        : 'Clouds',
                    'description' : 'broken clouds',
                    'icon'        : '04d',
                },
            ],
            'pop'        : 0.03,
        },
        {
            'dt'         : 1646316000,
            'temp'       : 281.94,
            'feels_like' : 278.49,
            'pressure'   : 1014,
            'humidity'   : 67,
            'dew_point'  : 276.16,
            'uvi'        : 1.49,
            'clouds'     : 52,
            'visibility' : 10000,
            'wind_speed' : 7.16,
            'wind_deg'   : 313,
            'wind_gust'  : 10.71,
            'weather'    : [
                {
                    'id'          : 803,
                    'main'        : 'Clouds',
                    'description' : 'broken clouds',
                    'icon'        : '04d',
                },
            ],
            'pop'        : 0.03,
        },
    ],
    'daily'           : [
        {
            'dt'         : 1646326800,
            'sunrise'    : 1646306882,
            'sunset'     : 1646347929,
            'moonrise'   : 1646309880,
            'moonset'    : 1646352120,
            'moon_phase' : 0.03,
            'temp'       : {
                'day'   : 281.63,
                'min'   : 271.72,
                'max'   : 282.21,
                'night' : 271.72,
                'eve'   : 277.99,
                'morn'  : 280.92,
            },
            'feels_like' : {
                'day'   : 277.83,
                'night' : 264.72,
                'eve'   : 273.35,
                'morn'  : 277.66,
            },
            'pressure'   : 1016,
            'humidity'   : 55,
            'dew_point'  : 273.12,
            'wind_speed' : 9.29,
            'wind_deg'   : 3,
            'wind_gust'  : 16.48,
            'weather'    : [
                {
                    'id'          : 500,
                    'main'        : 'Rain',
                    'description' : 'light rain',
                    'icon'        : '10d',
                },
            ],
            'clouds'     : 49,
            'pop'        : 0.25,
            'rain'       : 0.11,
            'uvi'        : 3.38,
        },
        {
            'dt'         : 1646326800,
            'sunrise'    : 1646306882,
            'sunset'     : 1646347929,
            'moonrise'   : 1646309880,
            'moonset'    : 1646352120,
            'moon_phase' : 0.03,
            'temp'       : {
                'day'   : 281.63,
                'min'   : 271.72,
                'max'   : 282.21,
                'night' : 271.72,
                'eve'   : 277.99,
                'morn'  : 280.92,
            },
            'feels_like' : {
                'day'   : 277.83,
                'night' : 264.72,
                'eve'   : 273.35,
                'morn'  : 277.66,
            },
            'pressure'   : 1016,
            'humidity'   : 55,
            'dew_point'  : 273.12,
            'wind_speed' : 9.29,
            'wind_deg'   : 3,
            'wind_gust'  : 16.48,
            'weather'    : [
                {
                    'id'          : 500,
                    'main'        : 'Rain',
                    'description' : 'light rain',
                    'icon'        : '10d',
                },
            ],
            'clouds'     : 49,
            'pop'        : 0.25,
            'rain'       : 0.11,
            'uvi'        : 3.38,
        },
    ],
};

const createStore = (preloadedState? : Record<string, any>) => configureStore({
    reducer : {
        settings : settingsReducer,
    },
    preloadedState,
});

const defaultStore = createStore();

type RootState = ReturnType<typeof defaultStore.getState>;

export const defaultState : RootState = {
    settings : {
        language         : 'en',
        darkMode         : false,
        autoLaunch       : false,
        refreshInterval  : 5 * 60 * 1000,
        timeFormat       : '12h',
        unit             : 'metric',
        locationProvider : 'here',
        weatherProvider  : 'openweathermap',
        forecastDays     : 4,
        forecastHours    : 8,
        forecastType     : 'humidity',
        places           : [ DEFAULT_LOCATION, ],
        selectedPlace    : DEFAULT_LOCATION,
        backdropDarken   : BACKGROUND_DARKEN,
        backdropBlur     : BACKGROUND_BLUR,
    },
};

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : PreloadedState<RootState>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => {
        const theme = createTheme({
            palette : {
                ...palette,
                mode : 'light',
            },
        });

        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <HashRouter>
                        {children}
                    </HashRouter>
                </ThemeProvider>
            </Provider>
        );
    },
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
