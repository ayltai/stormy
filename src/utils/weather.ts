export const getAlphaForPrecipIntensity = (intensity : number) => intensity === 0 ? 0 : intensity <= 1 ? 0.25 : intensity <= 2 ? 0.5 : intensity <= 4 ? 0.75 : 1;
