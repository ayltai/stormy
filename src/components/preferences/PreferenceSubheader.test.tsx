import React from 'react';

import { render, } from '../../utils/test';

import { PreferenceSubheader, } from './PreferenceSubheader';

describe('<PreferenceSubheader />', () => {
    it('renders correctly', () => {
        expect(render(<PreferenceSubheader title='dummy' />).asFragment()).toMatchSnapshot();
    });
});
