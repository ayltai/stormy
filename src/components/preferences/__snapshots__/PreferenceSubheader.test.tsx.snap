// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreferenceSubheader /> renders correctly 1`] = `
<DocumentFragment>
  <li
    class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-uob957-MuiListSubheader-root"
    role="heading"
  >
    <span
      aria-label=""
      class="MuiTypography-root MuiTypography-caption  css-1wkfms4-MuiTypography-root"
      data-mui-internal-clone-element="true"
    >
      dummy
    </span>
  </li>
</DocumentFragment>
`;
